framework/src/media/MediaPlayerImpl.cpp

```diff
@@ -617,12 +654,66 @@ void MediaPlayerImpl::notifyObserver(player_observer_command_t cmd, ...)
 			// Because data buffer would be released after this function returned.
 			mPlayerObserver->onPlaybackBufferDataReached(mPlayer, data, size);
 		} break;
+		case PLAYER_OBSERVER_COMMAND_ASYNC_PREPARED:
+			pow.enQueue(&MediaPlayerObserverInterface::onAsyncPrepared, mPlayerObserver, mPlayer, (player_error_t)va_arg(ap, int));
+			break;
 		}
 	}
 
 	va_end(ap);
 }
 
+void MediaPlayerImpl::notifyAsync(player_event_t event)
+{
+	LOG_STATE_INFO(mCurState);
+
+	if (mCurState != PLAYER_STATE_PREPARING) {
+		meddbg("%s Fail : invalid state\n", __func__);
+		LOG_STATE_DEBUG(mCurState);
+		return notifyObserver(PLAYER_OBSERVER_COMMAND_ASYNC_PREPARED, PLAYER_ERROR_INVALID_STATE);
+	}
+
+	switch (event) {
+	case PLAYER_EVENT_SOURCE_PREPARED: {
+		if (!mInputHandler.open()) {
+			meddbg("MediaPlayer prepare fail : input handler open fail\n");
+			return notifyObserver(PLAYER_OBSERVER_COMMAND_ASYNC_PREPARED, PLAYER_ERROR_FILE_OPEN_FAILED);
+		}
+
+		auto source = mInputHandler.getInputDataSource();
+		if (set_audio_stream_out(source->getChannels(), source->getSampleRate(),
+								 source->getPcmFormat()) != AUDIO_MANAGER_SUCCESS) {
+			meddbg("MediaPlayer prepare fail : set_audio_stream_out fail\n");
+			return notifyObserver(PLAYER_OBSERVER_COMMAND_ASYNC_PREPARED, PLAYER_ERROR_INTERNAL_OPERATION_FAILED);
+		}
+
+		mBufSize = get_user_output_frames_to_byte(get_output_frame_count());
+		if (mBufSize < 0) {
+			meddbg("MediaPlayer prepare fail : get_user_output_frames_to_byte fail\n");
+			return notifyObserver(PLAYER_OBSERVER_COMMAND_ASYNC_PREPARED, PLAYER_ERROR_INTERNAL_OPERATION_FAILED);
+		}
+
+		medvdbg("MediaPlayer mBuffer size : %d\n", mBufSize);
+
+		mBuffer = new unsigned char[mBufSize];
+		if (!mBuffer) {
+			meddbg("MediaPlayer prepare fail : mBuffer allocation fail\n");
+			if (get_errno() == ENOMEM) {
+				return notifyObserver(PLAYER_OBSERVER_COMMAND_ASYNC_PREPARED, PLAYER_ERROR_OUT_OF_MEMORY);
+			} else {
```

Comment:
here, else is not neccessary, please return directly.

Reply:
fixed, thanks